cub/cub/detail/rfa.cuh

```diff
@@ -0,0 +1,1258 @@
+    {
+      bins_host[i] = RFA_bins<ftype>::initialize_bins(i);
+    }
+
+    if (index >= MAXINDEX + MAXFOLD)
+    {
+      return static_cast<ftype>(bins_host[MAXINDEX + MAXFOLD - 1]);
+    }
+    if (index < 0)
+    {
+      return static_cast<ftype>(bins_host[0]);
+    }
+    return static_cast<ftype>(bins_host[index]);
+#endif
+  }
+
+  /// Get the bit representation of a float
+  __host__ __device__ static inline uint32_t& get_bits(float& x)
+  {
+    return *reinterpret_cast<uint32_t*>(&x);
+  }
+  /// Get the bit representation of a double
+  __host__ __device__ static inline uint64_t& get_bits(double& x)
+  {
+    return *reinterpret_cast<uint64_t*>(&x);
+  }
+  /// Get the bit representation of a const float
+  __host__ __device__ static inline uint32_t get_bits(const float& x)
+  {
+    return *reinterpret_cast<const uint32_t*>(&x);
+  }
+  /// Get the bit representation of a const double
+  __host__ __device__ static inline uint64_t get_bits(const double& x)
+  {
+    return *reinterpret_cast<const uint64_t*>(&x);
+  }
+
+  /// Return primary vector value const ref
+  __host__ __device__ __forceinline__ const ftype& primary(int i) const
+  {
+    if constexpr (FOLD <= MAX_JUMP)
+    {
+      switch (i)
+      {
+        case 0:
+          if constexpr (FOLD >= 1)
+          {
+            return data[0];
+          }
+        case 1:
+          if constexpr (FOLD >= 2)
+          {
+            return data[1];
+          }
+        case 2:
+          if constexpr (FOLD >= 3)
+          {
+            return data[2];
+          }
+        case 3:
+          if constexpr (FOLD >= 4)
+          {
+            return data[3];
+          }
+        case 4:
+          if constexpr (FOLD >= 5)
+          {
+            return data[4];
+          }
+        default:
+          return data[FOLD - 1];
+      }
+    }
+    else
+    {
+      return data[i];
+    }
+  }
+
+  /// Return carry vector value const ref
+  __host__ __device__ __forceinline__ const ftype& carry(int i) const
+  {
+    if constexpr (FOLD <= MAX_JUMP)
+    {
+      switch (i)
+      {
+        case 0:
+          if constexpr (FOLD >= 1)
+          {
+            return data[FOLD + 0];
+          }
+        case 1:
+          if constexpr (FOLD >= 2)
+          {
+            return data[FOLD + 1];
+          }
+        case 2:
+          if constexpr (FOLD >= 3)
+          {
+            return data[FOLD + 2];
+          }
+        case 3:
+          if constexpr (FOLD >= 4)
+          {
+            return data[FOLD + 3];
+          }
+        case 4:
+          if constexpr (FOLD >= 5)
+          {
+            return data[FOLD + 4];
+          }
+        default:
+          return data[2 * FOLD - 1];
+      }
+    }
+    else
+    {
+      return data[FOLD + i];
+    }
+  }
+
+  /// Return primary vector value ref
+  __host__ __device__ __forceinline__ ftype& primary(int i)
+  {
+    const auto& c = *this;
+    return const_cast<ftype&>(c.primary(i));
```

Comment:
I do not understand the purpose of the `c` temporary and the const cast. Could you elaborate why it is needed and not just:
```suggestion
    return primary(i);
```

Reply:
Won't that be a recursive call to itself?

Reply:
@miscco  I think here is your [answer](https://github.com/maddyscientist/reproducible_floating_sums/commit/2b3a5dc9b0dff9222edb38715069993ec2dcf090#diff-709208cd982e1ca086c637c561d56bc79208c6cdf08558123fad61210e6c13a5L166-R190)